src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs

```diff
@@ -9,295 +9,445 @@
 using System.Collections.Generic;
 using System;
 using Roslyn.Utilities;
+using Microsoft.CodeAnalysis.CSharp;
 
 namespace Microsoft.CodeAnalysis.CSharp
 {
     internal partial class LocalRewriter
     {
-        // PROTOTYPE(typeswitch): as a temporary hack while this code is in development, we
-        // only use the new translation machinery when this bool is set to true. If it is false
-        // then we use the transitional code which translates a pattern switch into a series of
-        // if-then-else statements. Ultimately we need the new translation to be used to generate
-        // switch IL instructions for ordinary old-style switch statements.
-        private static bool UseNewTranslation(BoundPatternSwitchStatement node) => false;
-
         public override BoundNode VisitPatternSwitchStatement(BoundPatternSwitchStatement node)
         {
-            // Until this is all implemented, we use a dumb series of if-then-else
-            // statements to translate the switch statement.
             _factory.Syntax = node.Syntax;
-            if (!UseNewTranslation(node)) return VisitPatternSwitchStatement_Ifchain(node);
-
-            var usedLabels = new HashSet<LabelSymbol>();
-            var usedTemps = new HashSet<LocalSymbol>();  // PROTOTYPE(typeswitch): worry about deterministic ordering
-            var result = ArrayBuilder<BoundStatement>.GetInstance();
-
-            // PROTOTYPE(typeswitch): do we need to do anything with node.ConstantTargetOpt, given that we
-            // have the decision tree? If not, can we remove it from the bound trees?
+            var pslr = new PatternSwitchLocalRewriter(this, node);
             var expression = VisitExpression(node.Expression);
-
-            if (node.DecisionTree.Expression != expression)
-            {
-                // Store the input expression into a temp
-                // PROTOTYPE(typeswitch): do we need to add the temp to the list of used temps?
-                result.Add(_factory.Assignment(node.DecisionTree.Expression, expression));
-            }
+            var result = ArrayBuilder<BoundStatement>.GetInstance();
 
             // output the decision tree part
-            LowerDecisionTree(node.DecisionTree, usedLabels, usedTemps, result);
-            result.Add(_factory.Goto(node.BreakLabel));
+            pslr.LowerDecisionTree(expression, node.DecisionTree, result);
+
+            // if the endpoint is reachable, we exit the switch
+            if (!node.DecisionTree.MatchIsComplete) result.Add(_factory.Goto(node.BreakLabel));
```

Comment:
I'll add scores of useless lines in this source file to be consistent with that.

---
In reply to: [70326276](https://github.com/dotnet/roslyn/pull/12416#discussion_r70326276) [](ancestors = 70326276)